pkg/engine2/operational_eval/vertex_resource_rule.go

```diff
@@ -0,0 +1,86 @@
+package operational_eval
+
+import (
+	"errors"
+	"fmt"
+
+	construct "github.com/klothoplatform/klotho/pkg/construct2"
+	"github.com/klothoplatform/klotho/pkg/engine2/operational_rule"
+	knowledgebase "github.com/klothoplatform/klotho/pkg/knowledge_base2"
+)
+
+type (
+	resourceRuleVertex struct {
+		Resource construct.ResourceId
+		Rule     knowledgebase.AdditionalRule
+	}
+)
+
+func (v resourceRuleVertex) Key() Key {
+	return Key{Ref: construct.PropertyRef{Resource: v.Resource}, RuleHash: v.Rule.Hash()}
```

Comment:
A little awkward to re-hash every call to `Key()` since we call it often.

Reply:
i can create the hash as a part of reading them in, so will change that